core/src/main/java/org/apache/druid/guice/GuiceInjectableValues.java

```diff
@@ -50,7 +59,21 @@ public Object findInjectableValue(
     // Currently we should only be dealing with `Key` instances, and anything more advanced should be handled with
     // great care
     if (valueId instanceof Key) {
-      return injector.getInstance((Key) valueId);
+      try {
+        return injector.getInstance((Key) valueId);
+      }
+      catch (ConfigurationException ce) {
+        // check if nullable annotation is present for this
+        if (nullables.get().contains((Key) valueId)) {
+          return null;
+        } else if (forProperty.getAnnotation(Nullable.class) != null) {
+          HashSet<Key> encounteredNullables = nullables.get();
+          encounteredNullables.add((Key) valueId);
```

Comment:
This is a non-thread-safe mutation of the HashSet.  The goal of the AtomicReference was to make thinds thread-safe.  You must create a brand new Set, add all of the old values, add the new value and then set the new reference on the AtomicReference.

Reply:
Addressed